### PR TITLE
Support local error type declaration

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Following is closure usage within error type is enabled with this PR.

```ballerina
final int k = 2;    
error<record {| int i = k; |}> e = error("hi");
```

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26652

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 1]
```diff
@@ -1287,6 +1287,11 @@ public void visit(BLangErrorType errorTypeNode) {
         BType detailType = Optional.ofNullable(errorTypeNode.detailType)
                 .map(bLangType -> resolveTypeNode(bLangType, env)).orElse(symTable.detailType);
 
+        if (errorTypeNode.isAnonymous) {
+            errorTypeNode.flagSet.add(Flag.PUBLIC);
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:1291]: Do we need PUBLIC also?

**Review by KRVPerera — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java (hunk 1 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java:1291]: We usually do this to handle the case where this local definition could be a part of a public construct.
`public function foo() returns error<{ int code; }> { }`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 2]
```diff
@@ -974,19 +974,20 @@ public BLangNode transform(ErrorTypeDescriptorNode errorTypeDescriptorNode) {
...
         }
 
         NonTerminalNode parent = errorTypeDescriptorNode.parent();
         boolean isDistinctError = parent.kind() == SyntaxKind.DISTINCT_TYPE_DESC;
         if (isDistinctError) {
             parent = parent.parent();
         }
-        if ((typeParam.isPresent() || isDistinctError) && parent.kind() != SyntaxKind.TYPE_DEFINITION) {
+
+        errorType.isAnonymous = checkIfAnonymous(errorTypeDescriptorNode);
```
Comment (sanjana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:986]: This causes a breaking change with #30678 due to being out of sync.

Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:986]: Will be addressed in #31084

**Pull request merged**